formats/packfile/decoder.go

```diff
@@ -87,39 +83,64 @@ func (d *Decoder) doDecode() error {
 		return err
 	}
 
-	if d.o == nil {
-		return d.readObjects(count)
+	_, isTxStorer := d.o.(core.ObjectStorerTx)
+	switch {
+	case d.o == nil:
+		return d.readObjects(int(count))
+	case isTxStorer:
+		return d.readObjectsWithObjectStorerTx(int(count))
+	default:
+		return d.readObjectsWithObjectStorer(int(count))
 	}
+}
 
-	if err := d.readObjects(count); err != nil {
-		if err := d.tx.Rollback(); err != nil {
-			return nil
+func (d *Decoder) readObjects(count int) error {
+	for i := 0; i < count; i++ {
+		if _, err := d.ReadObject(); err != nil {
+			return err
 		}
-
-		return err
 	}
 
-	return d.tx.Commit()
+	return nil
 }
 
-func (d *Decoder) readObjects(count uint32) error {
-	for i := 0; i < int(count); i++ {
+func (d *Decoder) readObjectsWithObjectStorer(count int) error {
```

Comment:
That's because you have another `readObjects` without a `s.o` set (the `core.ObjectStorer`), I think is clear enough being in the context (is a deep one)